### PR TITLE
Improve NEST GPU Utilization 4/N: 250x Faster Masking on GPU

### PR DESCRIPTION
**Pull request by MahmoudAshraf97**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

SSL training requires feature masking which uses python for loops on gpu hot path, this causes poor utilization since the masking function execution time grows linearly with batch size
This PR vectorizes the masking functions and gains 250x speedup using `block_size=48`, `masking_prob=0.01`, and `input_shape=[32,80,2000]`, 41ms down to 0.16ms, I did test higher batch sizes, but the utilization tanks for both implementations because the dataloader is the bottleneck at that stage, so I use grad accumulation for higher global batch sizes if needed

We get around 18% speedup E2E and 10-15% higher GPU utilization, I have verified that the masking distribution is the same in the overlapping path, and two training runs for 5k steps converged to the same loss

{1EE0C04C-C583-4032-8F76-355A10A3CBB7}

{6ABDEC86-9B75-4FE1-A876-3E9CFF0A54E8}

{C621F4E8-8E46-47B2-B0A8-5F74750A8E53}

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?
cc @stevehuang52 @pzelasko 
Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to #14086 #14089 #14234

### CODE REVIEW

**Comment by stevehuang52** [comment]
Thanks for the PR. We also have some optimization on masking and quantization in https://github.com/NVIDIA-NeMo/NeMo/tree/heh/ssl-v2 and the draft PR https://github.com/NVIDIA-NeMo/NeMo/pull/15589. Need to compare the two before deciding which one to merge.

**Comment by MahmoudAshraf97** [comment]
> Thanks for the PR. We also have some optimization on masking and quantization in https://github.com/NVIDIA-NeMo/NeMo/tree/heh/ssl-v2 and the draft PR #15589. Need to compare the two before deciding which one to merge.

I looked at the draft PR, I noticed that it still uses the old implementation core, where it basically uses python for loops so it will have the same issues, given that it does change the masking logic, we can use my implementation as a starting point and implement the new logic on top of the vectorized implementation

**Review by stevehuang52 — APPROVED** [review]
Looks good to me, thanks for the improvement~!

**Comment by stevehuang52** [comment]
/ok to test [df56f3d](https://github.com/NVIDIA-NeMo/NeMo/pull/15739/commits/df56f3d82326fdd28e96fe0cf1b18738ffe5a490)

**Comment by MahmoudAshraf97** [comment]
The two failing tests are not related to the code, please retry again or skip if unnecessary

**Comment by stevehuang52** [comment]
/ok to test [ac242b3](https://github.com/NVIDIA-NeMo/NeMo/pull/15739/commits/ac242b3cd6e37aa646a2ce94fb16bb697980e27a)

**Comment by stevehuang52** [comment]
/ok to test [5521ebf](https://github.com/NVIDIA-NeMo/NeMo/pull/15739/commits/5521ebf1529eb9c11aa7c8f1632b4353d40586a5)

**Comment by stevehuang52** [comment]
/ok to test [9bdbb9d](https://github.com/NVIDIA-NeMo/NeMo/pull/15739/commits/9bdbb9d5d9863a69163d35982571dab532cf3f45)

**Review by stevehuang52 — APPROVED** [review]
